### PR TITLE
Split core expression

### PR DESCRIPTION
**Pull request by mmatera**
@rocky, after all the afternoon-evening, `mathics.core.expression` has been split and tests seem to pass.

### CODE REVIEW

**Comment by rocky** [comment]
@mmatera thanks for doing this. I've run into that  regexp deprecation warnings before. `?u` is a flag for unicode and it can only appear once at the beginning of the regexp now.

**Comment by mmatera** [comment]
Something that still bothers me about this is that we need now to do imports inside methods, because some methods of `Expression` use `Symbol`, some methods of `Symbol` use `Expression`, and both use some of the `systemsymbols` (like `SymbolList` or `SymbolN`). Maybe we can find a way to avoid this in a further stage.

File: examples/symbolic_logic/gries_schneider/test_gs.py [hunk 1]
```diff
@@ -2,7 +2,9 @@
 # -*- coding: utf-8 -*-
 
 
-from mathics.core.expression import Expression, Integer, Rational, Symbol
+from mathics.core.expression import Expression
```
Comment (rocky) [inline examples/symbolic_logic/gries_schneider/test_gs.py:5]: flake8 says these three imports aren't used.

File: mathics/core/systemsymbols.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# cython: language_level=3
+# -*- coding: utf-8 -*-
+
+from mathics.core.symbols import (
```
Comment (rocky) [inline mathics/core/systemsymbols.py:4]: flake8 informs that none of these imports other than Symbol are used.

File: mathics/core/symbols.py [hunk 3]
```diff
@@ -0,0 +1,783 @@
...
+            if include_form:
+                expr = Expression(form, expr)
+            expr.unformatted = unformatted
+            return expr
+        finally:
+            evaluation.dec_recursion_depth()
+
+    def format(
+        self, evaluation, form, **kwargs
+    ) -> typing.Union["Expression", "Symbol"]:
```
Comment (rocky) [inline mathics/core/symbols.py:354]: I think `BaseExpression` is used, we don't need the Union or strings. And flake8 seems to be hapier.

File: mathics/core/symbols.py [hunk 4]
```diff
@@ -0,0 +1,783 @@
...
+            return self == rhs
+        return None
+
+    def has_changed(self, definitions):
+        return True
+
+    def sequences(self):
+        return None
+
+    def flatten_sequence(self, evaluation) -> "BaseExpression":
```
Comment (rocky) [inline mathics/core/symbols.py:126]: No quotes are needed for "BaseExpresion"

File: mathics/core/symbols.py [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+
+    def to_mpmath(self):
+        return None
+
+    def round_to_float(self, evaluation=None, permit_complex=False):
+        """
+        Try to round to python float. Return None if not possible.
+        """
+        from mathics.core.expression import Expression
+        from mathics.core.atoms import String, Number
```
Comment (rocky) [inline mathics/core/symbols.py:435]: flake8 informs that String is not used.

File: mathics/core/symbols.py [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+            value = self
+        elif isinstance(evaluation, sympy.core.numbers.NaN):
+            return None
+        else:
+            value = Expression(SymbolN, self).evaluate(evaluation)
+        if isinstance(value, Number):
+            value = value.round()
+            return value.get_float_value(permit_complex=permit_complex)
+
+    def __abs__(self) -> "Expression":
```
Comment (rocky) [inline mathics/core/symbols.py:447]: If BaseExpression is used no quotes are needed and flake8 is happier.

**Review by rocky — COMMENTED** [review]
@mmatera I want this to work so badly. I am seeing an old problem come up that I've never figured out previously. That there is some sort of weirdness in mpmath that prevents some modules from importing. 

When I run `mathics`  I see: 
```cannot import name 'Complex'
    Not able to load files-to-merge. Check your installation.
    mathics.builtin loads from /src/external-vcs/github/Mathics3/mathics-core/mathics/builtin/
```

However things still work: 

```
Mathics 4.0.1.dev0
on CPython 3.6.15 (default, Sep 21 2021, 05:10:37) 
using SymPy 1.8, mpmath 1.2.1, numpy 1.19.5

Copyright (C) 2011-2021 The Mathics Team.
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions.
See the documentation for the full license.

Quit by evaluating Quit[] or by pressing CONTROL-D.

In[1]:= 1+2
Out[1]= 3

In[2]:= 1I + 2I
Out[2]= 3 I
```

Is this what you see too? 

I will be investigating.

File: mathics/core/systemsymbols.py (hunk 2 above)
Comment (mmatera) [reply mathics/core/systemsymbols.py:4]: OK, but I want to have them available for all the modules that import `systemsymbols`... Maybe I should add a comment here...

**Comment by mmatera** [comment]
> @mmatera I want this to work so badly. I am seeing an old problem come up that I've never figured out previously. That there is some sort of weirdness in mpmath that prevents some modules from importing.
> 
> When I run `mathics` I see:
> 
> ```
>     Not able to load files-to-merge. Check your installation.
>     mathics.builtin loads from /src/external-vcs/github/Mathics3/mathics-core/mathics/builtin/
> ```
> 
> However things still work:
> 
> ```
> Mathics 4.0.1.dev0
> on CPython 3.6.15 (default, Sep 21 2021, 05:10:37) 
> using SymPy 1.8, mpmath 1.2.1, numpy 1.19.5
> 
> Copyright (C) 2011-2021 The Mathics Team.
> This program comes with ABSOLUTELY NO WARRANTY.
> This is free software, and you are welcome to redistribute it
> under certain conditions.
> See the documentation for the full license.
> 
> Quit by evaluating Quit[] or by pressing CONTROL-D.
> 
> In[1]:= 1+2
> Out[1]= 3
> 
> In[2]:= 1I + 2I
> Out[2]= 3 I
> ```
> 
> Is this what you see too?
> 
> I will be investigating.

No, I do not get this error:


```
(mathics3) mauricio@thinkpad:~/Projects/mathics-core$ mathics





Mathics 4.0.1.dev0
on CPython 3.9.1 (default, Dec 11 2020, 14:32:07) 
using SymPy 1.8, mpmath 1.2.1, numpy 1.20.1, cython 0.29.22

Copyright (C) 2011-2021 The Mathics Team.
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions.
See the documentation for the full license.

Quit by evaluating Quit[] or by pressing CONTROL-D.

In[1]:= 

```

File: mathics/core/symbols.py (hunk 4 above)
Comment (mmatera) [reply mathics/core/symbols.py:128]: flake8 complains if I remove the quotes...

**Review by rocky — APPROVED** [review]
LGTM - I have a number of additional changes, but this is large. So let's merge in and iterate. 

The other changes I have will be in a separate PR.

**Comment by rocky** [comment]
> > @mmatera I want this to work so badly. I am seeing an old problem come up that I've never figured out previously. That there is some sort of weirdness in mpmath that prevents some modules from importing.
> > When I run `mathics` I see:
> > ```
> >     Not able to load files-to-merge. Check your installation.
> >     mathics.builtin loads from /src/external-vcs/github/Mathics3/mathics-core/mathics/builtin/
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However things still work:
> > ```
> > Mathics 4.0.1.dev0
> > on CPython 3.6.15 (default, Sep 21 2021, 05:10:37) 
> > using SymPy 1.8, mpmath 1.2.1, numpy 1.19.5
> > 
> > Copyright (C) 2011-2021 The Mathics Team.
> > This program comes with ABSOLUTELY NO WARRANTY.
> > This is free software, and you are welcome to redistribute it
> > under certain conditions.
> > See the documentation for the full license.
> > 
> > Quit by evaluating Quit[] or by pressing CONTROL-D.
> > 
> > In[1]:= 1+2
> > Out[1]= 3
> > 
> > In[2]:= 1I + 2I
> > Out[2]= 3 I
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is this what you see too?
> > I will be investigating.
> 
> No, I do not get this error:
> 
> ```
> (mathics3) mauricio@thinkpad:~/Projects/mathics-core$ mathics
> 
> 
> 
> 
> 
> Mathics 4.0.1.dev0
> on CPython 3.9.1 (default, Dec 11 2020, 14:32:07) 
> using SymPy 1.8, mpmath 1.2.1, numpy 1.20.1, cython 0.29.22
> 
> Copyright (C) 2011-2021 The Mathics Team.
> This program comes with ABSOLUTELY NO WARRANTY.
> This is free software, and you are welcome to redistribute it
> under certain conditions.
> See the documentation for the full license.
> 
> Quit by evaluating Quit[] or by pressing CONTROL-D.
> 
> In[1]:= 
> ```

It was my fault - additional garbage on disk.

**Pull request merged**